pkgs/development/compilers/zig/generic.nix

```diff
@@ -0,0 +1,67 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, llvmPackages
+, libxml2
+, zlib
+, coreutils
+, ...
+}:
+
+hash: args:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "zig";
+
+  src = fetchFromGitHub {
+    owner = "ziglang";
+    repo = "zig";
+    rev = finalAttrs.version;
+    inherit hash;
+  };
+
+  nativeBuildInputs = [
+    cmake
+    llvmPackages.llvm.dev
+  ];
+
+  buildInputs = [
+    libxml2
+    zlib
+  ] ++ (with llvmPackages; [
+    libclang
+    lld
+    llvm
+  ]);
+
+  doInstallCheck = true;
+
+  # Zig's build looks at /usr/bin/env to find dynamic linking info. This doesn't
+  # work in Nix's sandbox. Use env from our coreutils instead.
+  postPatch = ''
+    substituteInPlace lib/std/zig/system/NativeTargetInfo.zig \
+      --replace "/usr/bin/env" "${coreutils}/bin/env"
+  '';
+
+  preBuild = ''
+    export ZIG_GLOBAL_CACHE_DIR=$(mktemp -d)
+  '';
+
+  installCheckPhase = ''
+    runHook preInstallCheck
+
+    $out/bin/zig test --cache-dir "$(mktemp -d)" -I $src/test $src/test/behavior.zig
```

Comment:
```suggestion
    $out/bin/zig test --cache-dir "$TMPDIR/cache-dir" -I $src/test $src/test/behavior.zig
```

Reply:
why are we using random names for temporary directories? `$TMPDIR/cache-dir` sounds like it can collide with another program easily

Reply:
- If `--cache-dir` is not set, then Zig will try to create one in `$HOME`.
- mktemp creates a random dir, introducing impurity
- `$TMPDIR` is set beforehand by Nix, so we are free to create directories inside it. No worries about collisions.

Reply:
`mktemp` seems to be more common in nixpkgs. The impurity shouldn't matter because if the temporary directory gets to the output it should already be a bug in the nix expression.

```console
$ rg '\$TMPDIR/' | wc -l
214

$ rg  '\$\(mktemp -d' | wc -l
517
```

Reply:
Strange. I have run ripgrep inside Emacs and it returned different results:

- [mktemp: 451](https://binshare.net/EpNeP37xKStOeikx9K3w)
- [tmpdir: 457](https://binshare.net/MNil3wsf4Ej3SzLQ5MqW)

Reply:
That was probably because I had a trailing slash after `$TMPDIR`, and your results included a bunch of `HOME=$TMPDIR`s, which, although I also don't agree with, feels a bit different because of the specialness of `$HOME`

Reply:
>  because of the specialness of `$HOME`

There are many `HOME=$(mktemp -d)` too.
Also the older Zig expressions did the same `HOME=$TMPDIR`, without the trailing slash.

Reply:
opted for a less generic name